File: android-native/app/src/main/java/com/example/kpkn/domain/exercises/ExerciseFilters.kt
package com.example.kpkn.domain.exercises

import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.MuscleRole
import com.example.kpkn.domain.training.VolumeCalculator

fun matchesMuscle(exercise: ExerciseMuscleInfo, anatomy: MuscleAnatomy): Boolean {
    return exercise.involvedMuscles.any {
        VolumeCalculator.normalizeCanonicalMuscleGroup(it.muscle, it.emphasis) == anatomy.canonicalName
            && it.role == MuscleRole.PRIMARY
    }
}

fun matchesMuscleHead(
    exercise: ExerciseMuscleInfo,
    anatomy: MuscleAnatomy,
    head: MuscleHead,
): Boolean {
    if (!matchesMuscle(exercise, anatomy)) return false
    return exercise.involvedMuscles.any { involvement ->
        val canonical = VolumeCalculator.normalizeCanonicalMuscleGroup(involvement.muscle, involvement.emphasis)
        if (canonical != anatomy.canonicalName) return@any false
        MuscleHeadResolution.matchesHead(anatomy.canonicalName, involvement.emphasis, head)
    }
}

File: ios-native/KPKNFit/KPKNFit/Domain/Exercises/ExerciseFilters.swift
import Foundation

func matchesMuscle(_ exercise: ExerciseMuscleInfo, anatomy: MuscleAnatomy) -> Bool {
    let involvements = exercise.involvedMuscles.filter {
        VolumeCalculator.normalizeCanonicalMuscleGroup($0.muscle, emphasis: $0.emphasis) == anatomy.canonicalName
    }
    if involvements.isEmpty { return false }

    if anatomy.canonicalName.lowercased() == "trapecio" {
        return involvements.contains { $0.role == .PRIMARY || $0.role == .SECONDARY }
    }
    return true
}

func matchesMuscleHead(
    _ exercise: ExerciseMuscleInfo,
    anatomy: MuscleAnatomy,
    head: MuscleHead
) -> Bool {
    if !matchesMuscle(exercise, anatomy: anatomy) { return false }
    return exercise.involvedMuscles.contains { involvement in
        let canonical = VolumeCalculator.normalizeCanonicalMuscleGroup(involvement.muscle, emphasis: involvement.emphasis)
        guard canonical == anatomy.canonicalName else { return false }
        let emphasisLower = (involvement.emphasis ?? "").lowercased().trimmingCharacters(in: .whitespaces)
        if let keyword = head.emphasisKeyword {
            return emphasisLower == keyword
        } else {
            return emphasisLower.isEmpty
        }
    }
}
